pkg/mcclient/mcclient.go

```diff
@@ -0,0 +1,158 @@
+	Cluster(name string) (client mcclient.Interface)
+	// ConfigForCluster returns rest config for given cluster.
+	ConfigForCluster(name string) (*rest.Config, error)
+	// CrossClusterListWatch returns cross-cluster ListWatch
+	CrossClusterListWatch(gv schema.GroupVersion, resource string, namespace string, fieldSelector fields.Selector) *listwatch.CrossClusterListerWatcher
+}
+
+type multiClusterClient struct {
+	configs    map[string]*rest.Config
+	clientsets map[string]*mcclient.Clientset
+	lock       sync.Mutex
+}
+
+func (mcc *multiClusterClient) Cluster(name string) mcclient.Interface {
+	mcc.lock.Lock()
+	defer mcc.lock.Unlock()
+	clientset, ok := mcc.clientsets[name]
+	if !ok {
+		//TODO change to ClusterOrDie and panic? return an error?
+		panic("invalid cluster name")
+	}
+	return clientset
+}
+
+func (mcc *multiClusterClient) ConfigForCluster(name string) (*rest.Config, error) {
+	mcc.lock.Lock()
+	defer mcc.lock.Unlock()
+	if _, ok := mcc.configs[name]; !ok {
+		//TODO get from server
+		return nil, errors.New("failed to get config for cluster")
+	}
+	return mcc.configs[name], nil
+}
+
+// CrossClusterListWatch NOT implemented. WIP
+func (mcc *multiClusterClient) CrossClusterListWatch(gv schema.GroupVersion, resource string, namespace string, fieldSelector fields.Selector) *listwatch.CrossClusterListerWatcher {
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fieldSelector.String()
+	}
+	clusterLW := make(map[string]*cache.ListWatch)
+	for cluster, config := range mcc.configs {
+		clusterLW[cluster] = listwatch.ClusterListWatch(config, gv, resource, namespace, optionsModifier)
+	}
+	return listwatch.NewCrossClusterListerWatcher(clusterLW)
+}
+
+var client *multiClusterClient
+
+// NewMultiCluster creates new multi-cluster client and starts collecting cluster configs
+func NewMultiCluster(ctx context.Context, managerConfig *rest.Config) (KubestellarClusterInterface, error) {
+	if client != nil {
+		return client, nil
+	}
+	client = &multiClusterClient{
+		configs:    make(map[string]*rest.Config),
+		clientsets: make(map[string]*mcclient.Clientset),
+		lock:       sync.Mutex{},
+	}
+	managerClientset, err := ksclientset.NewForConfig(managerConfig)
+	if err != nil {
+		return client, err
+	}
+
+	client.startClusterCollection(ctx, managerClientset)
+	return client, nil
+}
+
+func (mcc *multiClusterClient) startClusterCollection(ctx context.Context, managerClientset *ksclientset.Clientset) {
+	clusterInformerFactory := ksinformers.NewSharedScopedInformerFactory(managerClientset, 0, metav1.NamespaceAll)
+	clusterInformer := clusterInformerFactory.Edge().V1alpha1().LogicalClusters().Informer()
+
+	clusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			clusterInfo, ok := obj.(*v1alpha1.LogicalCluster)
+			if !ok {
+				log.Error("unexpected object type. expected LogicalCluster")
+				return
+			}
+			go mcc.handleAdd(clusterInfo)
```

Comment:
Because this directly uses goroutines rather than running the object references through a work queue, this can cause concurrent goroutines working on the same cluster.

Also, by the time the forked handleAdd starts working on a given cluster, it might already be deleted.  Similarly, by the time a forked handleDelete starts working on the cluster, it might have been added again.